### ISSUE TITLE
Heap profiling uses 16x more memory on ARM64, than it does on x64

### ISSUE THREAD

**Issue by madscientist** [jemalloc/jemalloc#2358]
We have a program that attempts to keep memory usage to a value specified by the user: it will fill its cache until the memory usage is reached then free memory.  If the system cannot release enough memory for a certain amount of time (default 5 minutes) it will kill itself.  To determine the amount of memory in use, we first use mallctlbymib() with "epoch", then we use mallctlbymib() with "stats.allocated".   We are using JEmalloc 5.3 on Linux x64 and this works fine: the system will run in a stable state for a very long time.

We are now porting to Linux ARM64, and on ARM we find that jemalloc is not able to free enough memory to stay under the limit.  For example if we set the memory limit to 2G on a 16G system it runs forever using about 2G of RAM on x64, but on ARM after a while no matter how much cache we try to free, we can't get "stats.allocated" down and we kill ourselves.

Just to note, we compile jemalloc with our code and link it statically.  We invoke `configure` with `--without-export --disable-doc --enable-prof` on both platforms, but no other options.

I've attached two sets of jemalloc stats output: one for our x64 run and the other for our ARM run.  Since they work at slightly different speeds and the x64 test runs to completion while the ARM test fails partway through, they are not fully apples-to-apples.  But you can see that on ARM the "mapped" and "retained" values are significantly larger than on x64.

I do notice that the page size on x64 is 4k while the page size on ARM is 64k.  But, should that matter?  Doesn't that just mean ARM reserves fewer pages and breaks the 64k up into more blocks?

Anyone with any insight as to why this is happening, please pass along!  I'm not sure where to go from here.

[last-jemalloc-stats.x86.txt](https://github.com/jemalloc/jemalloc/files/10119349/last-jemalloc-stats.x86.txt)
[last-jemalloc-stats.arm.txt](https://github.com/jemalloc/jemalloc/files/10119350/last-jemalloc-stats.arm.txt)

**Comment by madscientist** [comment]
We have done some investigation.  It seems that jemalloc is reserving a LOT more memory that doesn't appear in any profiling output, etc. on ARM, than it does on x64.  This memory is added to the stats.allocated but since it's maintained by jemalloc we can't affect it by releasing memory ourselves.

It looks like the culprit for reserving all this memory, is this function:
```
imalloc_sample(static_opts_t *sopts, dynamic_opts_t *dopts, tsd_t *tsd,
    size_t usize, szind_t ind)
```
It appears to be that because we enable profiling, and because the page size on ARM is 64k instead of 4k, this causes very large amounts of memory (16x the amount on x64) to be reserved for sampling.  I'm not 100% sure I have all the details correct on this.

Is this a bug, or just the way things work?  Is there something we can do to reduce the overhead on ARM, back to be closer to what it is on x64?

**Comment by interwq** [comment]
I'll try taking a more detailed look at this soon. This size class (the first in the `large` category) looks concerning on the ARM side:

```
large:          size ind    allocated      nmalloc (#/sec)      ndalloc (#/sec)    nrequests (#/sec)  curlextents
              262144  52    992739328       110766     151       106979     146       110766     151         3787
```

@madscientist would you be able to try disabling heap profiling and see if the memory usage is back to normal? It's likely still higher with the 64K page size, but shouldn't be dramatically so.

**Comment by madscientist** [comment]
Yes, those allocations are exactly the ones we were concerned about.  They are not being allocated by our code (we added something that logs all allocations >150,000 bytes and don't see more than a few, and the same on x64).  If you look at the smaller buffers on x64 you'll see a similar number of allocations there, that don't appear on the ARM system.

We believe it's due to the code in `imalloc_sample()` here:
```
		bumped_usize = sz_s2u(SC_LARGE_MINCLASS);
		ret = imalloc_no_sample(sopts, dopts, tsd, bumped_usize,
		    bumped_usize, ind_large);
```
On x64 the SC_LARGE_MINCLASS is 16k because LG_PAGE is 12, while on ARM SC_LARGE_MINCLASS is 256k because LG_PAGE is 16.

**Comment by interwq** [comment]
@madscientist the size bumping to 4x page size is indeed by design in the current implementation. That does seem too high with the larger page sizes (even though it might often be just VM). We are thinking about to reduce the size bumping to 1x page size instead, and possibly route it to a separate arena to facilitate reuse. It's a bit hard to promise when that will happen though (if you could get malloc_stats with profiling disabled to quantify the difference, it would certainly help us to prioritize).

In the meantime, some workaround you can consider on ARM: 
1) increase [lg_prof_sample](http://jemalloc.net/jemalloc.3.html#opt.lg_prof_sample), which will reduce the number of samples and the overhead. This may result in the heap profiles to be less representative.
2) or if acceptable, disable sampling when the page size is high for now.

**Comment by madscientist** [comment]
I'm working on getting a test run with sampling disabled.  Probably it won't be until tomorrow or Friday.

Unfortunately it's not acceptable to disable sampling.  These processes run for days / weeks / longer and if they die due to memory issues, the profiling information has been invaluable and indeed, often the only possible way we could figure out what happened (often we are not able to investigate core files, and if the kernel OOM killer gets involved there's no core anyway).

I can investigate the `lg_prof_sample` idea, but I'm not sure how this helps.  Won't this simply have the same problem, but more slowly?

Perhaps an issue is that we are using `stats.allocated` / `stats.resident` etc. to track our memory usage, and this is including not only memory that we (and our libraries) allocated but also the "extra" memory which is being used internally by jemalloc for things like sampling.  Is there any way we could retrieve from jemalloc only the primary memory usage, and/or discover the size of this "extra" memory so we could ignore it?

**Comment by madscientist** [comment]
Note I'm happy to apply local patches and changes if needed (obviously not the first option, but we already have a change we've made locally that allows us to dump profiling information to a memory buffer in addition to writing it to a file, so we have a process for maintaining local changes).

I can modify the code to change SC_LG_MINCLASS to a single page instead.  Unfortunately our testing resources are a bit thin due to other demands at the moment so my ability to just run experiments and try out different things is degraded.

It would help me if I can understand a bit about this allocation.  You mentioned it was done on purpose, but I'm not clear on why it's needed.  If we can get away with 16k on x64 why do we need 256k (or even a single 64k page) on ARM?  I'm concerned that even if we reduce the size to a single page it will still be a significant overhead on ARM.  Maybe I am just not understanding the idea behind this allocation.

**Comment by interwq** [comment]
> I can modify the code to change SC_LG_MINCLASS to a single page instead.

Unfortunately it would require a more involved change. Two reasons why sampled allocations are bumped to the large category (if they were not already) --
 
1) sampled allocations need some extra metadata, specifically in this union:
https://github.com/jemalloc/jemalloc/blob/70d4102f48dce2d5755e9139a15eeec606f97bff/include/jemalloc/internal/edata.h#L249

and making it large makes sure we only have a single allocation managed by this metadata (edata_t), so that the `e_slab_data` part is never used.

2) on deallocation, we need a quick way to detect sampled allocations since they need special treatment (to be un-sampled). Making it `large` give us the special alignment guarantee (PAGE aligned) so that we don't have to access metadata to check is sampled or not.

Both of these can be changed, the first one we'll need to change the assumptions made by the current impl (that all sampled are large). It will need some careful tweaking to make sure that we still only manage 1 allocation using the specific metadata. The second one I feel we still want the special alignment -- otherwise on every deallocation, the metadata needs to be accessed which would cause a significant slowdown of sized-delete. That's where the bump-to-page-size part comes in. Sorry that I don't think we have a quick and good workaround there.

> I can investigate the lg_prof_sample idea, but I'm not sure how this helps. Won't this simply have the same problem, but more slowly?

This depends on how many of the sampled allocations are "persistent" / long-lived / never goes away. Based on the stats you shared, there were ~3800 sampled allocations. Usually increasing the sampling interval should reduce that part accordingly, unless the workload tends to have the long-lived allocations sampled (so eventually all sampled allocations are long lived).

**Comment by dmn22-3ds** [comment]
I am a colleague of the mad scientist.

I don't think profiling less frequently will solve the problem. When the process has run for long enough, there will be a large number of objects that have been profiled and have not yet been deleted, so stats.allocated and stats.resident will still make it appear that we have run out of memory. Profiling less frequently will just make it run longer before crashing, I think.

We don't explicitly use arenas, so we would be perfectly happy with profiled (sampled) allocations being routed to a dedicated arena. Maybe that would allow you to manage your metadata without changing the size of the allocation.

**Comment by madscientist** [comment]
We have a run on ARM64 with profiling disabled; as we all probably expected it ran to completion with no OOM errors.  I've attached the stats generated near the end of the run if someone wants to examine them.

[last-jemalloc-stats-arm-noprof.txt](https://github.com/jemalloc/jemalloc/files/10158113/last-jemalloc-stats-arm-noprof.txt)

**Comment by madscientist** [comment]
Just wanted to mention: while RedHat has decided to change the kernel's page size on ARM back to 4k starting with RHEL9 (which is what all other Linux vendors do also on their ARM systems AFAICT), IBM still uses 64k pages on Power, even with RHEL9 and as far as I can tell they don't have any intention of changing this in the future.

So it would be great if someone, someday could work out a solution to allow profiling to be enabled on larger kernel page sizes without enormous overhead.  Just sayin' :)

**Comment by YuSongHao** [comment]
> We have a run on ARM64 with profiling disabled; as we all probably expected it ran to completion with no OOM errors. I've attached the stats generated near the end of the run if someone wants to examine them.
> 
> [last-jemalloc-stats-arm-noprof.txt](https://github.com/jemalloc/jemalloc/files/10158113/last-jemalloc-stats-arm-noprof.txt)

Hi @madscientist ，I meet the same issue in JEmalloc 5.3.0. The difference is the memory consumption did not decrease after disable profiling(--disable-prof). I just want to know where I did something wrong, can you post your configuration?

**Comment by madscientist** [comment]
We just invoke `configure` without the `--enable-prof` option on ARM.  Disabled is the default.  But I think unless you've really dug into the problem and proven it's related to profiling, that should not be your first assumption: there are lots of ways that your memory might grow that have nothing to do with profiling.

I've been tracking the changes above that will hopefully help this issue but haven't had time to try them on our systems yet.  Not sure if there's a new jemalloc release containing these coming any time soon or if I'll have to try to backport the changes.

**Comment by madscientist** [comment]
This issue is still open: is the expectation that the changes pushed above last July would address it?  Is it waiting for me to verify or something before closing the issue?  I'm unfamiliar with the process :)

**Comment by interwq** [comment]
@madscientist it was improved by #2459 -- the overhead per sampled small allocation lowered to 1 page (from 4 pages previously). You can try the current dev branch. The most recent commit mentioned in #2060 would be a safe bet.

To be clear, there is still some overhead, as the page size is often configured as 16K or 64K on ARM, vs x64's page size of 4K. The 1 page per sample cost can still be high with the larger pages, but improving it further won't be easy, and likely will come with some speed tradeoffs.

**Comment by madscientist** [comment]
Thank you for those details @interwq .  Just to be clear, that change to reduce the overhead is in effect for all platforms, so systems with 4k pages also use only 4k per sample instead of 16k which is used in the 5.3 release?  If so that's nice.

I assume another way to reduce the overhead is to increase the `opt.lg_prof_sample` so we take fewer samples?  Can this be modified dynamically at any time?  Do I have to do something like disable/re-enable profiling?

**Comment by interwq** [comment]
> that change to reduce the overhead is in effect for all platforms, so systems with 4k pages also use only 4k per sample instead of 16k which is used in the 5.3 release?

Right. You might see a small memory reduction on the 4K platforms as well.

> I assume another way to reduce the overhead is to increase the opt.lg_prof_sample so we take fewer samples? Can this be modified dynamically at any time?

Yes opt.lg_prof_sample can be used to lower the number of samples and therefore the overhead as well. I'd suggest take a look at the current number of samples in the generated profiles, and if it looks high then consider lowering it accordingly.

> Can this be modified dynamically at any time?

Unfortunately no. The reason is that it would complicate the extrapolation used when reporting the final results. It's not impossible to account for different sampling intervals, but certainly adds more bits into it.

**Comment by madscientist** [comment]
> > Can this be modified dynamically at any time?

> Unfortunately no.

So just to be clear you're saying I need to set it up in `malloc_conf` correct?

**Comment by interwq** [comment]
@madscientist right. Set it in `malloc_conf` (but it's not changeable via `mallctl`s after the binary starts).
